Modules/_multiprocessing/posixshmem.c

```diff
@@ -0,0 +1,724 @@
+        self->name = (char *)PyMem_Malloc(strlen(temp_name) + 1);
+        if (self->name)
+            strcpy(self->name, temp_name);
+        else {
+            PyErr_SetString(PyExc_MemoryError, "Out of memory");
+            goto error_return;
+        }
+    }
+    else {
+        // (name != None) ==> use name supplied by the caller. It was
+        // already converted to C by convert_name_param().
+        self->name = name.name;
+
+        DPRINTF("calling shm_open, name=%s, flags=0x%x, mode=0%o\n",
+                    self->name, flags, (int)self->mode);
+        self->fd = shm_open(self->name, flags, (mode_t)self->mode);
+    }
+
+    DPRINTF("shm fd = %d\n", self->fd);
+
+    if (-1 == self->fd) {
+        self->fd = 0;
+        switch (errno) {
+            case EACCES:
+                PyErr_Format(pPermissionsException,
+                                "No permission to %s this segment",
+                                (flags & O_TRUNC) ? "truncate" : "access"
+                                );
+            break;
+
+            case EEXIST:
+                PyErr_SetString(pExistentialException,
+                                "Shared memory with the specified name already exists");
+            break;
+
+            case ENOENT:
+                PyErr_SetString(pExistentialException,
+                                "No shared memory exists with the specified name");
+            break;
+
+            case EINVAL:
+                PyErr_SetString(PyExc_ValueError, "Invalid parameter(s)");
+            break;
+
+            case EMFILE:
+                PyErr_SetString(PyExc_OSError,
+                                 "This process already has the maximum number of files open");
+            break;
+
+            case ENFILE:
+                PyErr_SetString(PyExc_OSError,
+                                 "The system limit on the total number of open files has been reached");
+            break;
+
+            case ENAMETOOLONG:
+                PyErr_SetString(PyExc_ValueError,
+                                 "The name is too long");
+            break;
+
+            default:
+                PyErr_SetFromErrno(PyExc_OSError);
+            break;
+        }
+
+        goto error_return;
+    }
+    else {
+        if (size) {
+            DPRINTF("calling ftruncate, fd = %d, size = %ld\n", self->fd, size);
+            if (-1 == ftruncate(self->fd, (off_t)size)) {
+                // The code below will raise a Python error. Since that error
+                // is raised during __init__(), it will look to the caller
+                // as if object creation failed entirely. Here I clean up
+                // the system object I just created.
+                close(self->fd);
+                shm_unlink(self->name);
+
+                // ftruncate can return a ton of different errors, but most
+                // are not relevant or are extremely unlikely.
+                switch (errno) {
```

Comment:
Same here: why not use PyErr_SetFromErrno?

Reply:
Removed in GH-11816.